### PR TITLE
Pull and extract the content package set in distro.prop

### PR DESCRIPTION
**Pull request by nravilla**
## Description of what I changed
Modified classes to download and extract and move content packages (just the backend content)

## Issue I worked on
https://openmrs.atlassian.net/browse/O3-3640

## Checklist: I completed these to help reviewers :)
- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

- [x] Acceptance tests (setup and build-distro was tested and it works

- [x] I ran `mvn clean install` right before creating this pull request and
  added all formatting changes to my commit.

- [x] All existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

File: sdk-commons/src/main/java/org/openmrs/maven/plugins/model/BaseSdkProperties.java [hunk 1]
```diff
@@ -1,333 +1,372 @@
...
+		for (Object keyObject : getAllKeys()) {
+			String key = keyObject.toString();
+			String artifactType = getArtifactType(key);
+			if (artifactType.equals(TYPE_CONTENT)) {
+				String artifactId = key.substring(TYPE_CONTENT.length() + 1);
+				System.out.println("n\n");
+				System.out.println(artifactId);
+				System.out.println(getParam(key));
+				System.out.println(Artifact.GROUP_CONTENT);
+
```
Comment (mherman22) [inline sdk-commons/src/main/java/org/openmrs/maven/plugins/model/BaseSdkProperties.java:180]: i think we can remove these `System.out.println`s during clean up after all the reviews

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ModuleInstaller.java
Comment (mherman22) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ModuleInstaller.java:1]: was this change intended?

**Review by mherman22 — COMMENTED** [review]
Thanks @nravilla a few comments as we wait for @ibacher's review

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/Setup.java [hunk 2]
```diff
@@ -319,20 +324,29 @@ private void setJdk(Server server) throws MojoExecutionException {
...
 			File owasDir = new File(server.getServerDirectory(), "owa");
 			owasDir.mkdir();
 			downloadOWAs(server.getServerDirectory(), distroProperties, owasDir);
 		}
 	}
-
+	
+	private void installContents(Server server, DistroProperties distroProperties) throws MojoExecutionException {
+		if (distroProperties != null) {
+			File tempContentDir = new File(server.getServerDirectory(), "temp-content");
```
Comment (ibacher) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/Setup.java:341]: Same comments here as before

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+		this.session = session;
+		this.mavenProject = mavenProject;
+		this.pluginManager = pluginManager;
+	}
+	
+	public void downloadContent(File configurationDir, File tempContentDir, Artifact contentArtifact,
+	        ModuleInstaller moduleInstaller) throws MojoExecutionException {
+		
+		String artifactId = contentArtifact.getArtifactId();
+		File contentDir = new File(tempContentDir, artifactId);
```
Comment (ibacher) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java:36]:
```suggestion
		File contentPackageDir = new File(contentDir, artifactId);
```

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+		moduleInstaller.installUnpackModule(contentArtifact, contentDir.getAbsolutePath());
+		moveBackendConfig(contentDir, configurationDir, artifactId);
+	}
+	
+	private void moveBackendConfig(File inputDir, File configurationDir, String artifactId) throws MojoExecutionException {
+		
+		// Check if inputDir and outputDir are valid directories
+		if (!inputDir.isDirectory() || !configurationDir.isDirectory()) {
+			throw new MojoExecutionException("Both inputDir and outputDir must be valid directories.");
+		}
```
Comment (ibacher) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java:47]: As a non-public method, we kind of already know that `inputDir` is a valid directory and the check for `configurationDir` probably should go at the start of the public method, since that's where any client application will be calling from (it also saves us from doing work if it's unnecessary).

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+		
+		// List all directories in the inputDir
+		File backendConfigs = new File(inputDir, "configs" + File.separator + "backend_config");
```
Comment (ibacher) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java:50]:
```suggestion
		File backendConfigs = Paths.get(inputDir, "configs", "backend_config").toFile();
```

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+		
+		if (backendConfigs == null || backendConfigs.listFiles().length == 0) {
+			throw new MojoExecutionException("No directories to process under content configuration directories");
+		}
```
Comment (ibacher) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java:54]: There is no way that `backendConfigs` could be `null` here.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+		
+		for (File backendConfig : backendConfigs.listFiles()) {
```
Comment (ibacher) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java:56]:
```suggestion
		for (File configDir : backendConfigs.listFiles()) {
```

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+			if (!matchingConfigurationDir.exists() && !matchingConfigurationDir.isDirectory()) {
+				//this folder is from content zip
+				artifactDir = matchingConfigurationDir;
+			} else {
+				// Create a new folder with the artifactId under the matching output directory
+				artifactDir = new File(matchingConfigurationDir, artifactId);
+				if (!artifactDir.exists()) {
+					artifactDir.mkdirs();
+				}
+			}
```
Comment (ibacher) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java:71]: Am I reading the logic here correctly? It says "if matchingConfigurationDir does not exist and it is not a directory, set artifactDir to matchingConfigurationDir, otherwise create a new subfolder"?

Assuming I'm reading that correctly:

1. Under what circumstance can the `if` branch be true?
2. We always want to ensure that we are writing to a subfolder of the configuration directory, right?

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+			
+			// Copy the content from inputDir to the newly created artifactDir
+			copyDirectory(backendConfig, artifactDir);
```
Comment (ibacher) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ContentHelper.java:74]: `FileUtils.copyDirectory()` can be used here.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ModuleInstaller.java [hunk 10]
```diff
@@ -24,98 +10,113 @@
...
-                        groupId(SDKConstants.DEPENDENCY_PLUGIN_GROUP_ID),
-                        artifactId(SDKConstants.DEPENDENCY_PLUGIN_ARTIFACT_ID),
-                        version(SDKConstants.DEPENDENCY_PLUGIN_VERSION)
-                ),
-                goal(goal),
-                configuration(configuration.toArray(new Element[0])),
-                executionEnvironment(mavenProject, mavenSession, pluginManager)
-        );
-    }
+
```
Comment (ibacher) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ModuleInstaller.java:31]: I don't love all the reformatting in this PR. It makes it very hard to review and know if anything has changed and if so, what.

Comment (ibacher) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/ModuleInstaller.java:31]: Please don't commit files you haven't touched other than reformatting. If there are any changes here, can you please use GitHub comments to indicate what those are?

**Review by ibacher — CHANGES_REQUESTED** [review]
Thanks @nravilla. Because of the number of formatting changes, this is a really hard PR to review. I've left a few pointers on things that stuck out to me, but I'll need more time to review the remaining files in any detail.

**Comment by nravilla** [comment]
@ibacher @mherman22 Unable to resolve the formatting issues, will close this and create a new branch
